### PR TITLE
Handle audio, video, image, 3D model + integrate backend for Marker based experience

### LINKED ISSUE #32: Change uploaded files properties (see description)
For the MVP, we should change, if possible, the sizes and rotation of GLTF files uploaded (also 2D video and 2D images).

* for GLTF, we have to change sizes if too big, and define a standard size
* same for 2D videos
* same for 2D images

and for every of them, position should be set at "0,0,0" while rotation should be made so the content is perpendicular to the marker.

This changes are not to be done, obviously, on the source file the user uploaded: we should take the sizes, rotation and position of the given content, and knowing the output we want, we should cable the scale, rotation and position properties according to this data.

E.G.

We set '100' as default size for Images.

If the users loads a 200px images, we should call the backend with scale parameters data, in this way:

`scale: "0.5, 0.5, 0.5"`

### LINKED ISSUE #36: Create marker based experience
On marker based page, connect the frontend with backend in order to create html marker based experience

### LINKED ISSUE #37: Render video when uploaded on preview
When user uploaded a video, in both marker based / location based page, a preview should be showed. For the video, we can embed it

### LINKED ISSUE #38: Render image when uploaded on preview
When user uploaded an image, in both marker based / location based page, a preview should be showed. We can show the image on the preview box, for example

### LINKED ISSUE #39: Render audio when uploaded on preview
When user uploaded an audio asset, in both marker based / location based page, a preview should be showed. We can embed a basic audio player, from HTML 5 specifics, on the preview box

### LINKED ISSUE #40: Render 3D model on preview when uploaded
When user uploaded a 3D model, in both marker based / location based page, a preview should be showed. We can embed an aframe scene (or three.js viewer) on the preview box and show the 3D model. Zoom/rotating enabled would be great but not needed as MVP

### PR DESCRIPTION
**Pull request by mingir2**
Ok this became a much larger Pull Request now. I had to change this main post. 
So, this actually:

Closes #36  
Closes #37 
Closes #38 
Closes #39 
Closes #40 
Closes #32 

Things left to fix:

- [x] Understand why the scaling of 3D model is giving problems, and also find a starting scale (maybe we can find a ratio between modelFrame box and the original 3D model width, and find the starting scale ratio)

- [x] Understand how we can handle .gltf files. .glb files are ok as they have all resources in one file, instead gltf MAY come with different files, that now are not handled because we upload just .gltf

- [x] Handle errors (what happens if I select video, for example, and load an image? or what happens If I try to load something without selecting the type?)

- [x] Fix style (size of upload boxes, check on responsive, missing cursor: pointer on buttons, etc.)

Please write other missing things on the checklist above.

### CODE REVIEW

**Comment by mingir2** [comment]
I'm working on error handling as well.

**Comment by nicolocarpignoli** [comment]
@mingir2 had a chat with @RomanKlimov that told me he has started to work on integration of backed/frontend and also developed a bit of this feature of render image. He might help you, i'm sure we can take the best of your efforts to solve this :)

**Comment by RomanKlimov** [comment]
I used @mingir2 code for this feature and did the same for marker preview, there is still some css work, pushed it all on feature/marker_based_exp_for_image :)

**Comment by mingir2** [comment]
@RomanKlimov nice work! I have some CSS fixes and some other stuff. I'll merge your changes to my branch.

By the way, is there a particular reason you put a script below the page-footer?

**Comment by mingir2** [comment]
I moved the "free" <script> in the marker.html to marker.js. I'm guessing that was done due to the event listener binds, so I tagged the marker.js with defer.

Some other changes:
- Fix CSS align for the cross mark and file name
- Use CDN for fetching studo-backend api
- Fix handleImageUnload to handle other contents using id attribute
- Clean up zip function via chaining.
- Rename handleImageUnload to handleUnload.

For the html creation part, I haven't look in detail, but it looks like you might be generating a marker for the wrong image file. The first image that's uploaded needs to become a marker and the second image/3d/video/audio will be the AR content displayed/placed on top of the marker.

**Comment by nicolocarpignoli** [comment]
@mingir2 I updated the main post to keep track of every issue related to this one
@faace added problem for 3D models on the checklist. I have added 3D built in animation running on loop, if available on the model.

File: pages/marker.js [hunk 1]
```diff
@@ -181,22 +181,25 @@ const supportedFileMap = {
...
-    if (supportedFile) {
-        if (file.size > supportedFile.maxSize) errorMessage = 'exceed max size ' + supportedFile.maxSizeText;
-        else if (type === '3d') { // cannot get the file.type
-            let fileName = file.name.split('.');
-            let ext = fileName[fileName.length - 1];
-            if (supportedFile.types.indexOf(ext) < 0) errorMessage = 'error content type';
+const isSupportedFileAndSize = (type, file, id) => {
+    const supportedFile = supportedFileMap[type];
+    const previewError = document.getElementById(id)
+    previewError.innerHTML = ""
```
Comment (mingir2) [inline pages/marker.js:187]: This line resets the error statement.

**Comment by nicolocarpignoli** [comment]
I changed general spacing a bit. Tell me if you like it more. @fcor changed again XD But seems better to me, before this change, too few elements too much sparse on the screen, not they're more near between them.

**Comment by fcor** [comment]
Looks good! Good job @mingir2 @RomanKlimov @faace!

> Handle errors (what happens if I select video, for example, and load an image? or what happens If I try to load something without selecting the type?)

On this one, we could change `accept `property based on user's selection to only accept certain type of files. What do you think? @nicolocarpignoli

**Comment by faace** [comment]
@fcor We have changed `accept` property when the type is image/audio/video except the 3d model. It seems the 3d model type is not supported.

**Comment by nicolocarpignoli** [comment]
we should use a custom check for 3d model type, as now it fails type checking

**Comment by faace** [comment]
> we should use a custom check for 3d model type, as now it fails type checking

We have a check by the suffix of gltf and glb for 3d model type.

**Comment by nicolocarpignoli** [comment]
@faace can you update us about last commits?

**Comment by faace** [comment]
When image/audio/video type is selected, the filter will be set to the corresponding type, so users can only upload the corresponding files. If 3d type is selected. uses may upload files of any type. But we have a check of the suffix. If uses don't select a type. no file can be uploaded successfully.

**Comment by nicolocarpignoli** [comment]
@mingir2 @RomanKlimov can you help me test the error handling, upload of glb, upload of gltf, upload of zip?

After that, I will refine style a bit and we can close this huge PR  :)

**Comment by mingir2** [comment]
@nicolocarpignoli I've did some tests with models from [three.js](https://github.com/mrdoob/three.js/tree/master/examples/models/gltf/CesiumManl). 

- Error handling for filetype works, but it's currently hidden due to CSS conflict with main.css. error class is in both main.css and style.css and main.css sets visibility to hidden for error CSS class. This issue can be solved by either toggling visibility in marker.js or cleaning up CSS classes, so that they don't clash.
- We don't have the error handling for whether the upload was successful or not. If I upload a zipped gltf with an image file missing, then it will fail silently and display errors on the console. 
- Also, I tried uploading glb file, downloading package, unzip package, and then re-upload the asset to check whether it worked. It failed to load with ```
SyntaxError: Unexpected token e in JSON at position 0
    at JSON.parse (<anonymous>)
    at FileReader.reader.onloadend``` Seems like zipping process need to be revisited.

@faace It would be good to separate 3rd party libraries either through CDN or a dedicated folder. Also, feel free to create more js files to organize functions, especially since some of them can be used in the location page.

**Comment by mingir2** [comment]
I did a lot of refactoring work, so the implantations themselves are the same as before. I did this so that we can work on one small parts at a time without having to know every little detail. I still don't fully grasp everything that's happening on the page, but no one should need to.

One change worth mentioning here would be that the <file-selector> now modifies <input id="content-file"> "accept" parameters directly rather than in marker.js.

**Comment by nicolocarpignoli** [comment]
there's a bug:

- try to load an image while you have selected GLTF 3D model
- error shows, okay
- select now image as file to be uploaded
- try now to upload an image
- nothing happens

**Comment by faace** [comment]
> there's a problem:
> 
> * try to load an image while you have selected GLTF 3D model
> * error shows, okay
> * select now image as file to be uploaded
> * try now to upload an image
> * nothing happens
> 
> @mingir2 @faace

This problem is fixed in the latest submit.

**Comment by nicolocarpignoli** [comment]
@faace thank you. Two last things to fix now, before merging:

- I can't upload MP3 files, always gives me error even with file under the maximum allowed size
- The 'file-holder' components, both for marker and content, should be clickable and activate the upload function. Now they're not doing anything because there is no callback attached with the click event

**Comment by nicolocarpignoli** [comment]
we can merge this. thank you everybody for the great work!

**Pull request merged**